lib/src/dbus_ws_read_buffer.dart

```diff
@@ -0,0 +1,386 @@
+    Map<String,dynamic> jsonResult = {};
+
+    var type;
+    var serial;
+
+    if (_responseData.startsWith("invoke:")) {
+      // print("> readMessage ${_responseData}");
+      jsonResult = json.decode(_responseData.substring("invoke:".length));
+      serial = jsonResult['id']['serial'];
+
+      if (jsonResult['result']!=null) {
+        type = DBusMessageType.methodReturn;
+      } else if (jsonResult['error']!=null) {
+        type = DBusMessageType.error;
+      }
+      // print("> readMessage type: $type, jr: $jsonResult");
+    } else if (_responseData.startsWith("addMatch:")) {
+      if (jsonResult['result']!=null) {
+        type = DBusMessageType.methodReturn;
+      } else if (jsonResult['error']!=null) {
+        type = DBusMessageType.error;
+      }
+      serial = jsonResult['id'];
+      // print("> readMessage (addMatch) type: $type jr: ${_responseData}");
+    } else if (_responseData.startsWith("signal:")) {
+      type = DBusMessageType.signal;
+      jsonResult = json.decode(_responseData.substring("signal:".length));
+      serial = jsonResult['serial'];
+      // print("> readMessage type $type, jr: $jsonResult");
+    } 
+
+    // var type = {
+    //   1: DBusMessageType.methodCall,
+    //   2: DBusMessageType.methodReturn,
+    //   3: DBusMessageType.error,
+    //   4: DBusMessageType.signal
+    // }[readDBusByte()!.value];
+
+    if (type == null) {
+      throw 'Invalid type received';
+    }
+    var flags = <DBusMessageFlag>{};
+    if (jsonResult['flags']!=null) {
+      var flagsValue = jsonResult['flags'];
+      if (flagsValue & 0x01 != 0) {
+        flags.add(DBusMessageFlag.noReplyExpected);
+      }
+      if (flagsValue & 0x02 != 0) {
+        flags.add(DBusMessageFlag.noAutoStart);
+      }
+      if (flagsValue & 0x04 != 0) {
+        flags.add(DBusMessageFlag.allowInteractiveAuthorization);
+      }
+    }
+
+    DBusSignature? signature;
+    DBusObjectPath? path;
+    DBusInterfaceName? interface;
+    DBusMemberName? member;
+    DBusErrorName? errorName;
+    int? replySerial;
+    DBusBusName? destination;
+    DBusBusName? sender;
+    var fdCount = 0;
+
+    if (jsonResult['path']!=null) {
+      path = DBusObjectPath(jsonResult['path']);
+    }
+    if (jsonResult['interface']!=null) {
+      interface = DBusInterfaceName(jsonResult['interface']);
+    }
+    if (jsonResult['member']!=null) {
+      member = DBusMemberName(jsonResult['member']);
+    }
+    if (jsonResult['errorName']!=null) {
+      errorName = DBusErrorName(jsonResult['errorName']);
+    }
+    if (jsonResult['destination']!=null) {
+      destination = DBusBusName(jsonResult['destination']);
+    }
+    if (jsonResult['sender']!=null) {
+      sender = DBusBusName(jsonResult['sender']);
+      if (!(sender.value == 'org.freedesktop.DBus' || sender.isUnique)) {
+        throw 'Sender contains non-unique bus name';
+      }
+    }
+
+    var values = <DBusValue>[  ];
+    var values_in_json  = null;
```

Comment:
there are a couple of whitespace issues here and there (like here, two spaces before '=') - that could probably be improved with 'dart format'

Reply:
done